File: ShortestDistance.java
// Time Complexity : The time complexity is O(n) where n is the length of the array
// Space Complexity : Te space complexity is O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

class Solution {
    public int shortestDistance(String[] wordsDict, String word1, String word2) {

        int res = Integer.MAX_VALUE;
        int w1 = -1;
        int w2 = -1;

        for(int i=0;i<wordsDict.length;i++){
            String cur = wordsDict[i];

            //index of first word
            if(word1.equals(cur)){
                w1 = i;
            }

            //index of second word
            if(word2.equals(cur)){
                w2 = i;
            }

            // calculate the distance when we found both the words
            if(w1 >= 0 && w2 >= 0){
                res = Math.min(res,Math.abs(w2-w1));
            }
        }
        return res;
    }
}

File: shortestDistance.js
// Time Complexity : O(n) where n is length of wordsDict
// Space Complexity : O(1)

/**
 * @param {string[]} wordsDict
 * @param {string} word1
 * @param {string} word2
 * @return {number}
 */

var shortestDistance = function(wordsDict, word1, word2) {
    let p1 = -1;
    let p2 = -1;
    let minDist = Infinity
    for(let i = 0; i < wordsDict.length; i++) {
        if(wordsDict[i] === word1) p1=i;
        if(wordsDict[i] === word2) p2=i;
        if(p1 !== -1 && p2 !== -1) {
            minDist = Math.min(minDist, Math.abs(p2-p1));
        }
    }
    return minDist;
};